### PR TITLE
Adding function for arbitrary virtual image generation

### PR DESCRIPTION
**Pull request by cophus**
Copying method used in annular image generation

### CODE REVIEW

**Comment by cophus** [comment]
I placed an example notebook into slack, but the gist of the new function is something like:

mask_inner = py4DSTEM.process.utils.make_circular_mask((dataset.Q_Nx,dataset.Q_Ny), (qx0, qy0), det_DF[0])
mask_outer = py4DSTEM.process.utils.make_circular_mask((dataset.Q_Nx,dataset.Q_Ny), (qx0, qy0), det_DF[1])
mask = mask_outer & ~mask_inner

virtual_image = py4DSTEM.process.virtualimage.get_virtualimage(
    dataset,
    mask)

mask can be any boolean array - this function also uses the square subslicing method for a speed boost.

File: py4DSTEM/process/virtualimage/virtualimage.py [hunk 1]
```diff
@@ -85,7 +85,36 @@ def get_virtualimage_ann(datacube, x0, y0, Ri, Ro, verbose=True):
...
+    x = np.where(np.max(mask, axis=1))
+    y = np.where(np.max(mask, axis=0))
+    xmin, xmax = np.min(x), np.max(x)
+    ymin, ymax = np.min(y), np.max(y)
+    mask_sub = mask[xmin:xmax,ymin:ymax]
+
+    # Generate virtual image
+    virtual_image = np.zeros((datacube.R_Nx, datacube.R_Ny))
+    for rx,ry in tqdmnd(datacube.R_Nx, datacube.R_Ny, disable=not verbose):
+        virtual_image[rx,ry] = np.sum(datacube.data[rx,ry,xmin:xmax,ymin:ymax]*mask_sub)
```
Comment (sezelt) [inline py4DSTEM/process/virtualimage/virtualimage.py:116]: On my machine, this is 13% faster for the annular mask in the example notebook if you replace the multiplication by boolean with slicing by a boolean, i.e.
```
virtual_image[rx,ry] = np.sum(datacube.data[rx,ry,xmin:xmax,ymin:ymax][mask_sub])
```

File: py4DSTEM/process/virtualimage/virtualimage.py [hunk 2]
```diff
@@ -85,7 +85,36 @@ def get_virtualimage_ann(datacube, x0, y0, Ri, Ro, verbose=True):
     return virtual_image
 
 
+def get_virtualimage(
+    datacube, 
+    mask, 
+    verbose=True,
+    return_mask_coords=False,
```
Comment (sezelt) [inline py4DSTEM/process/virtualimage/virtualimage.py:92]: This is unused?

**Review by sezelt — CHANGES_REQUESTED** [review]
Looks great!
There is a small tweak, commented below, that will give a further performance boost (thanks to @PhilippPelz for the suggestion!).
Also, can you update the docstring to reflect the new arguments?

File: py4DSTEM/process/virtualimage/virtualimage.py (hunk 2 above)
Comment (cophus) [reply py4DSTEM/process/virtualimage/virtualimage.py:92]: Oops a leftover from the previous version!

File: py4DSTEM/process/virtualimage/virtualimage.py (hunk 1 above)
Comment (cophus) [reply py4DSTEM/process/virtualimage/virtualimage.py:116]: I always forget you can do this with numpy arrays - added!

Comment (cophus) [reply py4DSTEM/process/virtualimage/virtualimage.py]: Docstring updated for verbose too

**Pull request merged**